fem/lor.hpp

```diff
@@ -0,0 +1,287 @@
+
+   /// Resets the integration rules of the integrators of @a a to their original
+   /// values (after temporarily changing them for %LOR assembly).
+   void ResetIntegrationRules(GetIntegratorsFn get_integrators);
+
+   static inline int absdof(int i) { return i < 0 ? -1-i : i; }
+
+protected:
+   enum FESpaceType { H1, ND, RT, L2, INVALID };
+
+   FiniteElementSpace &fes_ho;
+   Mesh *mesh;
+   FiniteElementCollection *fec;
+   FiniteElementSpace *fes;
+   BilinearForm *a;
+   OperatorHandle A;
+   mutable Array<int> perm;
+   bool nonconforming = false;
+
+   /// Constructs the local DOF (ldof) permutation. In parallel this is used as
+   /// an intermediate step in computing the DOF permutation (see
+   /// ConstructDofPermutation and GetDofPermutation).
+   void ConstructLocalDofPermutation(Array<int> &perm_) const;
+
+   /// Construct the permutation that maps %LOR DOFs to high-order DOFs. See
+   /// GetDofPermutation.
+   void ConstructDofPermutation() const;
+
+   /// Sets up the prolongation and restriction operators required for
+   /// nonconforming spaces.
+   void SetupNonconforming();
+
+   /// Returns the type of finite element space: H1, ND, RT or L2.
+   FESpaceType GetFESpaceType() const;
+
+   /// Returns the order of the %LOR space. 1 for H1 or ND, 0 for L2 or RT.
+   int GetLOROrder() const;
+
+   LORBase(FiniteElementSpace &fes_ho_);
+
+public:
+   /// Returns the assembled %LOR system.
+   const OperatorHandle &GetAssembledSystem() const;
+
+   /// Assembles the %LOR system.
+   void AssembleSystem(BilinearForm &a_ho, const Array<int> &ess_dofs);
+
+   /// @brief Returns the permutation that maps %LOR DOFs to high-order DOFs.
+   ///
+   /// This permutation is constructed the first time it is requested, and then
+   /// is cached. For H1 and L2 finite element spaces (or for nonconforming
+   /// spaces) this is the identity. In these cases, RequiresDofPermutation will
+   /// return false. However, if the DOF permutation is requested, an identity
+   /// permutation will be built and returned.
+   ///
+   /// For vector finite element spaces (ND and RT), the DOF permutation is
+   /// nontrivial. Returns an array @a perm such that, given an index @a i of a
+   /// %LOR dof, @a perm[i] is the index of the corresponding HO dof.
+   const Array<int> &GetDofPermutation() const;
+
+   /// Returns true if the %LOR spaces requires a DOF permutation (if the
+   /// corresponding %LOR and HO DOFs are numbered differently), false
+   /// otherwise. Note: permutations are not required in the case of
+   /// nonconforming spaces, since the DOF numbering is incorporated into the
+   /// prolongation operators.
+   bool RequiresDofPermutation() const;
+
+   /// Returns the low-order refined finite element space.
+   FiniteElementSpace &GetFESpace() const { return *fes; }
+
+   ~LORBase();
+};
+
+/// Create and assemble a low-order refined version of a BilinearForm.
+class LOR : public LORBase
```

Comment:
I don't like the name `LOR` because I think it will be obscure to users and hard to google. I would prefer `LowOrderRefined`, or maybe `LORDiscretization` or `LORSystem`.

Reply:
I would also prefer a more descriptive name like `LowOrderRefined`.

Reply:
I changed it to `LORDiscretization` (which goes well with the related class, `LORSolver`). `LowOrderRefinedDiscretization` is probably too verbose.